Feature/working shared env

ion-oset opened the pull request:
Fixes #1.

The problem, as noted in the commit that does the fix, is in the internals of `poetry install`. To determine if a directory is package, Poetry [scans the directory and all its subdirectories](https://github.com/python-poetry/poetry-core/blob/1.2.0/src/poetry/core/masonry/utils/package_include.py#L83) looking for`.py` file. If it finds even one it considers the directory to be a package.

The source that does this uses `path.glob("**/*")` where `path` is a `pathlib.Path`.  If it had used `glob.glob("**/*", recurse = True)` it would work, but `Path.glob` doesn't traverse symlinks so it fails. This isn't really a bug _per se_. Poetry has no reason to expect that top-level package directories are all symbolic links.

The fix is to add a Python file to the top-level folder, so I added an `__init__.py`. This is a little perverse, because it means `src/vtp` in this project is no longer a namespace package. If this project was meant to be installed and imported itself this wouldn't work, but this project is just a shell for build purposes. All the other packages it is wrapping are still namespace packages, so they can share the package top-level.

ion-oset committed:
Clean up indentation of '.gitmodules'.

ion-oset committed:
Allow 'poetry install' to work with the symlinks in 'src/vtp'.

Unintuitively this is done by adding '__init__.py' which makes 'src/vtp' in
this project *not* be a namespace package. This is needed because 'poetry
install' is not following symlinks due to an implementation quirk.

Fixes #1.

windoverwater commented:
One minor comment - git naturally puts tabs in the .git/config file and I would imagine by extension all config files written by git, such as .gitmodules.  Is it incorrect to try to force them into spaces?

https://stackoverflow.com/questions/40091541/can-git-be-told-to-use-spaces-in-gitconfig

https://stackoverflow.com/questions/32186600/how-to-make-git-config-to-use-spaces-instead-of-tabs

windoverwater commented:
Side note - because this branch is based on a main commit in the past (conflicts with the current main) including commits which include submodule changes, and since this branch thus references non existent commits in the VTP-web-api submodule at this point, and since the .gitmodule commit above seems not to be critical, and since I want to use this now, I am just going to cherry-pick this change into main and push that.  Hopefully that is ok.

windoverwater commented:
FYI - for some reason TBD, when I try this and attempt to run the fastapi server, it fails with the following.  When I try to run the uvicorn server from just the VTP-web-api pyproject.toml file, the uvicorn server runs as desired.

However, in both environments (the VTP-dev-env and VoteTrackerPlus pyproject.toml files), the VoteTrackerPlus executables seem to work (the same).

[sandy] VTP-dev-env % poetry shell                                                                                             
Spawning shell within /Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10
Restored session: Wed Apr  5 13:29:13 EDT 2023
[sandy] VTP-dev-env % emulate bash -c '. /Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/bin/activate'
(vtpdevenv-py3.10) [sandy] VTP-dev-env % cd VTP-web-api/src/vtp/web/api                                                        
(vtpdevenv-py3.10) [sandy] api % uvicorn main:app                                                                              
Traceback (most recent call last):
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/bin/uvicorn", line 8, in <module>
    sys.exit(main())
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/main.py", line 404, in main
    run(
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/main.py", line 569, in run
    server.run()
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/server.py", line 67, in serve
    config.load()
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/config.py", line 477, in load
    self.loaded_app = import_from_string(self.app)
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/importer.py", line 24, in import_from_string
    raise exc from None
  File "/Users/sandy/Library/Caches/pypoetry/virtualenvs/vtpdevenv-05N0F1ji-py3.10/lib/python3.10/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/sandy/Library/CloudStorage/Dropbox/repos/VTP-dev-env/VTP-web-api/src/vtp/web/api/./main.py", line 8, in <module>
    import vtp.web.api.data
ModuleNotFoundError: No module named 'vtp.web'

ion-oset committed:
Fix broken import of 'vtp.web.api'.

There was no import path to 'vtp.web.api': the symlink was one level too high
('vtp.api').

ion-oset commented:
Fixed. The problem is what the exception says:

> ModuleNotFoundError: No module named 'vtp.web'

There was no path `src/vtp/web` in this repository because the symlink was `vtp.api` instead of `vtp/web/api`. Note that this was just a minor mistake in the base branch.

ion-oset commented:
> One minor comment - git naturally puts tabs in the .git/config file and I would imagine by extension all config files written by git, such as .gitmodules. Is it incorrect to try to force them into spaces?

Spaces work in `.git/config` and `.gitconfig`, it's just that Git itself always uses tabs. You can use tabs if you prefer. The change I made was to stop mixing tabs and spaces, but I use spaces by default so that's what my editor changed to.

ion-oset commented:
> since I want to use this now, I am just going to cherry-pick this change into main and push that. Hopefully that is ok.

Whatever you prefer. Just let me know since I didn't realize you did that and pushed the other fix here.

windoverwater commented:
@ion-oset As far as I know/can tell, after the cherry-pick of your changes into main everything is working ok.  So I am going to close this PR to keep things simple.
